heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,547 @@
+   * @param unit the time unit of the timeout argument
+   * @return true MesosFramework got registered,
+   * and false if the waiting time elapsed before the count reached zero
+   */
+  public boolean waitForRegistered(long timeout, TimeUnit unit) {
+    try {
+      if (this.registeredLatch.await(timeout, unit)) {
+        return true;
+      }
+    } catch (InterruptedException e) {
+      LOG.severe("Failed to wait for mesos framework got registered");
+      return false;
+    }
+
+    return false;
+  }
+
+  /**
+   * Get the FrameworkID registered with Mesos Master
+   *
+   * @return the FrameworkID
+   */
+  public Protos.FrameworkID getFrameworkId() {
+    return frameworkId;
+  }
+
+
+  @Override
+  public void registered(SchedulerDriver schedulerDriver,
+                         Protos.FrameworkID frameworkID,
+                         Protos.MasterInfo masterInfo) {
+    LOG.info("Registered! ID = " + frameworkID.getValue());
+    this.driver = schedulerDriver;
+    this.frameworkId = frameworkID;
+
+    // Release the latch
+    this.registeredLatch.countDown();
+  }
+
+  @Override
+  public void reregistered(SchedulerDriver schedulerDriver, Protos.MasterInfo masterInfo) {
+    LOG.info("Re-registered!");
+  }
+
+  @Override
+  public void resourceOffers(SchedulerDriver schedulerDriver, List<Protos.Offer> offers) {
+    LOG.fine("Received Resource Offers: " + offers.toString());
+
+    Map<Protos.Offer, TaskResources> offerResources = new HashMap<>();
+    // Convert the offers into TasksResources, an utils class.
+    for (Protos.Offer offer : offers) {
+      offerResources.put(offer, TaskResources.apply(offer, Context.role(heronConfig)));
+    }
+
+    // Generate launchable tasks
+    List<LaunchableTask> tasksToLaunch = generateLaunchableTasks(offerResources);
+
+    // Launch tasks
+    launchTasks(tasksToLaunch);
+
+    // Decline unused offers
+    LOG.fine("Declining unused offers.");
+    Set<String> usedOffers = new HashSet<>();
+    for (LaunchableTask task : tasksToLaunch) {
+      usedOffers.add(task.offer.getId().getValue());
+    }
+    for (Protos.Offer offer : offers) {
+      if (!usedOffers.contains(offer.getId().getValue())) {
+        schedulerDriver.declineOffer(offer.getId());
+      }
+    }
+  }
+
+  @Override
+  public void offerRescinded(SchedulerDriver schedulerDriver, Protos.OfferID offerID) {
+    LOG.info("Offer Rescinded!");
+  }
+
+  @Override
+  public void statusUpdate(SchedulerDriver schedulerDriver, Protos.TaskStatus taskStatus) {
+    LOG.info(String.format("Received status update [%s]", taskStatus));
+    if (isTerminated) {
+      LOG.info("The framework terminated. Ignoring any status update.");
+    }
+
+    String taskId = taskStatus.getTaskId().getValue();
+
+    BaseContainer container = containersInfo.get(TaskUtils.getContainerIndexForTaskId(taskId));
+
+    if (container == null) {
+      LOG.info("Received a status update from unknown container. ");
```

Comment:
Updated to log at WARNING level. 
Added a comment in the source code:
```
// TODO(mfu): Don't know what should be done here. Simply ignore this status update for now.
```
I am hesitate to kill the task here since it is also possible to kill tasks belonging to other jobs.
This case should rarely happen.